ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,482 @@
+//! Known bits analysis.
+//!
+//! Data-flow analysis to gain information about bits.
+//! We use similar ideas to PyPy:
+//! https://pypy.org/posts/2024/08/toy-knownbits.html
+//!
+
+use crate::compile::{
+    j2::{
+        hir::{And, Const, ConstKind, Inst, InstIdx, InstT, Or, Ty},
+        opt::{
+            BlockLikeT,
+            fullopt::{CommitInstOpt, OptOutcome, PassOpt, PassT},
+        },
+    },
+    jitc_yk::arbbitint::ArbBitInt,
+};
+use index_vec::IndexVec;
+
+/// Known-bits analysis.
+pub(super) struct KnownBits {
+    known_bits: IndexVec<InstIdx, Option<KnownBitValue>>,
+    pending_commit: Option<KnownBitValue>,
+}
+
+impl PassT for KnownBits {
+    fn feed(&mut self, opt: &mut PassOpt, inst: Inst) -> OptOutcome {
+        self.pending_commit = None;
+        match inst {
+            Inst::And(x) => self.opt_and(opt, x),
+            Inst::Const(x) => self.opt_const(x),
+            Inst::Or(x) => self.opt_or(opt, x),
+            _ => OptOutcome::Rewritten(inst),
+        }
+    }
+
+    fn inst_committed(&mut self, _opt: &CommitInstOpt, _iidx: InstIdx, _inst: &Inst) {
+        self.known_bits.push(self.pending_commit.clone());
+        assert_eq!(_iidx.index(), self.known_bits.len() - 1);
+    }
+}
+
+impl KnownBits {
+    /// Create an empty known bits analysis object.
+    pub(super) fn new() -> Self {
+        KnownBits {
+            known_bits: IndexVec::new(),
+            pending_commit: None,
+        }
+    }
+
+    /// Returns what we known about the bits of `iidx`.
+    fn as_knownbits(&self, opt: &PassOpt, iidx: InstIdx) -> KnownBitValue {
+        self.known_bits[iidx].clone().unwrap_or_else(|| {
+            KnownBitValue::unknown(match opt.inst(iidx).ty(opt) {
+                // The first two are currently unimplemented.
+                Ty::Func(_) => 0,
+                Ty::Void => 0,
+                ty => ty.bitw(),
+            })
+        })
+    }
+
+    fn set_knownbits(&mut self, bits: KnownBitValue) {
+        self.pending_commit = Some(bits);
+    }
+
+    /// the `and` operation adds new information in the form of set zero.
+    /// E.g. `unknown & (~1)` zeroes the least significant bit.
+    /// We thus define information gain as the knowledge of a set zero.
+    fn opt_and(&mut self, opt: &mut PassOpt, mut inst: And) -> OptOutcome {
+        inst.canonicalise(opt);
+        let And { tyidx, lhs, rhs } = inst;
+        let lhs_b = self.as_knownbits(opt, lhs);
+        let rhs_b = self.as_knownbits(opt, rhs);
+        let res = lhs_b.and(&rhs_b);
+        self.set_knownbits(res.clone());
+
+        // If we know the output's bits, emit that.
+        if res.all_known() {
+            return OptOutcome::Rewritten(Inst::Const(Const {
+                tyidx,
+                kind: ConstKind::Int(res.as_int()),
+            }));
+        }
+
+        // If no new information (set zeroes) was gained, that means this
+        // op is useless.
+        if rhs_b.all_known()
+            && (rhs_b
```

Comment:
I'm surprised `cargo fmt` hasn't removed the brackets. If it doesn't, we should remove them, as "over bracketing" isn't really needed in Rust like it is in C.